update configuration for Travis testing

reneeotten opened the pull request:
The commit in this PR does the following:

* remove testing for Python 3.3, which reached EOL

* test for current versions of SciPy/NumPy for all Python versions
This will install the latest version of scipy/numpy available in conda for a certain Python version. Currently, it uses numpy 1.13.3 and scipy 1.0.0 (PY27, PY35 and PY36) and numpy 1.11.3 and scipy 0.18.1 fo PY34.

* test for minimum required SciPy/NumPy for Python 2.7 and latest Python 3 version
To limit the number of builds I think it should be sufficient to test the minimum required version of numpy/scipy with PY27 and -ideally- the latest PY3 release. For PY27 we can do both builds with the minimum scipy/numpy versions, but for PY36 the requirements can only be satisfied for numpy. Therefore, the minimum scipy version is tested with PY34.

newville commented:
@reneeotten  Thanks!  I agree, though have some suggested tweaks.  

I'm +1 on dropping testing of Python 3.3. I'd even say we could drop 3.4 (it's hard to imagine people stuck on 3.3 or 3.4 and unable to upgrade to 3.5).   Let's just hurry up and get to Python 3.6+ only already!

For scipy versions, the biggest thing is that lmfit requires scipy > 0.15.   I think the policy should be (approximately) that we may stop testing and supporting version of scipy that are more than 2 years old.  

Currently, this would mean that we expect to support scipy 0.16.   If we fudge that a bit, we could say scipy 0.17 (released Jan 2016), which would mean we could rely on `least_squares()` being available.   I'd say that this is a fine expectation for the next release of lmfit, even if it slightly violates the 2-year policy (we should start planning the next release and also 1.0).

So, I think that means we should probably test with both scipy 0.17 and 1.0.0 for all of PY27, PY35, and PY36 and that we can start relying on `least_squares()` being available. 

Does that all seem reasonable to you (and anyone else:  @andyfaff @Tillsten)?

reneeotten commented:
I am favor of explicitly supporting/testing for old versions and would be happy to require scipy 0.17 and then perhaps also for numpy go to 0.10 (released in October 2015). For Python versions we could just follow the [status of Python branches](https://docs.python.org/devguide/#status-of-python-branches), which would still include testing for PY34. We could then test two different scenarios ```latest``` and ```minimum```, which would install the minimum required versions for scipy/numpy.

Anyway, the main issue with Travis and using ```conda``` is that not all packages are available for every Python version. Even if we would go to the above mentioned scipy/numpy version we can still not test for PY36. Perhaps it's possible to use ```pip``` with the ```--only-binary``` option to avoid compilation from source, but I haven't checked if this actually works. 

Let's come to a consensus first and then I'll update the PR.

newville commented:
On Sun, Nov 12, 2017 at 1:02 PM, Renee Otten <notifications@github.com>
wrote:

> I am favor of explicitly supporting/testing for old versions and would be
> happy to require scipy 0.17 and then perhaps also for numpy go to 0.10
> (released in October 2015). For Python versions we could just follow the status
> of Python branches
> <https://docs.python.org/devguide/#status-of-python-branches>, which
> would still include testing for PY34. We could then test two different
> scenarios latest and minimum, which would install the minimum required
> versions for scipy/numpy.
>
OK, yes, I agree with that approach to handling Python versions.  I think
requiring scipy 0.17 or higher for the next release is also fine.


> Anyway, the main issue with Travis and using conda is that not all
> packages are available for every Python version. Even if we would go to the
> above mentioned scipy/numpy version we can still not test for PY36. Perhaps
> it's possible to use pip with the --only-binary option to avoid
> compilation from source, but I haven't checked if this actually works.
>

Yes, this is definitely the challenge for writing .travis.yml.   I think
it's acceptable to do "best available" here.  If conda PY36 only goes back
to scipy version X.Y, that's fine.

Let's come to a consensus first and then I'll update the PR.
>
OK. I don't think we have to kill ourselves on the testing matrix. I think
most people using lmfit are going to be aware of the potential for version
conflicts between libraries.  As long as we're testing multiple
combinations spanning the likely ranges of versions, I think we can accpt
the risk that we might miss an occasional, subtle conflict.

reneeotten commented:
work in progress.... not ready to merge yet!

newville commented:
@reneeotten  OK, and thanks. FWIW, it seems like "consensus" is between you and me. I feel like I gave an opinion and didn't really try out any of the combinations.  You're doing the work, so I think I will agree with whatever you think is best.

reneeotten commented:
@newville this should be good to go. It incorporates your suggestions and in addition tests with a "minimal" installation: only the minimum versions of packages listed in ```requirements.txt``` are installed. 
Bumped the minimum version to scipy>=0.17 and numpy>=1.10; updated the documentation and removed the ```HAS_LEAST_SQUARES``` check.

Lets first merge #430 and then I will rebase this PR so that it merges cleanly.

newville commented:
@reneeotten.  OK, sounds like a fine plan.  I think this and #440 should be
orthogonal, so merge order ought to not matter much.  Either way, we can
make it work....

On Nov 19, 2017 10:35, "Renee Otten" <notifications@github.com> wrote:

> @newville <https://github.com/newville> this should be good to go. It
> incorporates your suggestions and in addition tests with a "minimal"
> installation: only the minimum versions of packages listed in
> requirements.txt are installed.
> Bumped the minimum version to scipy>=0.17 and numpy>=1.10; updated the
> documentation and removed the HAS_LEAST_SQUARES check.
>
> Lets first merge #430 <https://github.com/lmfit/lmfit-py/pull/430> and
> then I will rebase this PR so that it merges cleanly.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/lmfit/lmfit-py/pull/443#issuecomment-345530007>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AASke8OkW_WABBKLy9jKP6yXYJI-2PNHks5s4Fi6gaJpZM4Qar-8>
> .
>

newville commented:
@reneeotten This looks like it will merge without conflict.   Do you agree this is ready to merge?

reneeotten commented:
@newville  - no, it will merge cleanly, but then still some of the changes you made (i.e., always installing ```dill``) are in there as well... I will update it and push shortly, then you can merge.

reneeotten commented:
@newville ready-to-go - thanks!

newville commented:
@reneeotten Great! merging now.
